app/Jobs/GenerateSimilarityIndex.php

```diff
@@ -0,0 +1,215 @@
+    public $tries = 2;
+
+    /**
+     * Ignore this job if the image does not exist any more.
+     *
+     * @var bool
+     */
+    protected $deleteWhenMissingModels = true;
+
+
+    /**
+     * The volume to process
+     *
+     * @var Volume
+     */
+    public $volume;
+
+    /**
+     * The hash values according to the images in volume
+     *
+     * @var array
+     */
+    protected $hashValues;
+
+    /**
+     * The similarity values according to the images in volume
+     *
+     * @var array
+     */
+    protected $similarityIndices;
+
+    /**
+     * Create a new job instance.
+     *
+     * @param Image $image The image to generate process.
+     *
+     * @return void
+     */
+    public function __construct(Volume $volume)
+    {
+        $this->volume = $volume;
+        $this->hashValues = [];
+        $this->similarityIndices = [];
+
+    }
+
+    /**
+     * Execute the job.
+     *
+     * @return void
+     * @throws Exception
+     */
+    public function handle()
+    {
+        $collection = $this->volume->images;
+        if ($collection->count() !== 1) {
+            if ($collection->whereNull('hash')->isEmpty()) {
+                $imageHashArray = $collection->pluck('hash', 'id');
+                $similarityIndexArray = $this->createSimilarityIndexArray($imageHashArray);
+                if (!is_null($similarityIndexArray)) {
+                    foreach ($this->volume->images as $image) {
+                        $id = $image->id;
+                        $similarityIndex = $similarityIndexArray[$id];
+                        $image->similarityIndex = $similarityIndex;
+                        $image->save();
+                    }
+                }
+            }
+        } else {
+            foreach ($this->volume->images as $image) {
+                $image->similarityIndex = 0;
+                $image->save();
+            }
+        }
+
+
+
+
+
+
+
+    }
+
+    /**
+     * Execute the SimilarityIndexGenerator script for the Volume
+     * @throws Exception
+     */
+    public function createSimilarityIndexArray($imageHashArray)
+    {
+        try {
+            $script = config('biigle.similarity_index_generator');
+            $outputPath = $this->getOutputJsonPath($this->volume);
+            $inputPath = $this->createInputJson($this->volume, $imageHashArray);
+            $output = $this->python("{$script} {$inputPath} {$outputPath}");
+            $similarityIndexArray = json_decode(File::get($outputPath), true);
+      } catch (Exception $e) {
+            $input = File::get($inputPath);
+            $similarityIndexArray = null;
+            throw new Exception("Input: {$input}\n" . $e->getMessage());
+        } finally {
+            if (isset($inputPath)) {
+                $this->maybeDeleteFile($inputPath);
+            }
+
+            if (isset($outputPath)) {
+                $this->maybeDeleteFile($outputPath);
+            }
+        }
+        return $similarityIndexArray;
```

Comment:
Another tip: Make `return` statements stand out more clearly with a newline (unless they are the only line in a block).

```suggestion

        return $similarityIndexArray;
```